parachain/primitives/ethereum/src/ethashproof.rs

```diff
@@ -0,0 +1,353 @@
+        assert!(max > 0);
+        EthashCache {
+            max_capacity: max,
+            caches_by_epoch: BTreeMap::new(),
+            recently_accessed_epochs: Vec::with_capacity(max),
+            cache_gen_fn: Self::get_cache_for_epoch,
+        }
+    }
+
+    /// For tests to override the cache data generator
+    pub fn with_generator(max: usize, cache_gen_fn: fn(usize) -> Vec<u8>) -> EthashCache {
+        let mut cache = EthashCache::new(max);
+        cache.cache_gen_fn = cache_gen_fn;
+        cache
+    }
+
+    pub fn get(&mut self, epoch: u64, timestamp: u64) -> &Vec<u8> {
+        if self.caches_by_epoch.contains_key(&epoch) {
+            let (ref mut t, _e) = self.recently_accessed_epochs
+                .iter_mut()
+                .find(|&&mut pair| pair.1 == epoch)
+                .unwrap();
+            *t = timestamp;
+        } else {
+            if self.recently_accessed_epochs.len() == self.max_capacity {
+                let (ref mut t, ref mut e) = self.recently_accessed_epochs.first_mut().unwrap();
+                self.caches_by_epoch.remove(e);
+                *t = timestamp;
+                *e = epoch;
+            } else {
+                self.recently_accessed_epochs.push((timestamp, epoch));
+            }
+            let cache_gen_fn = self.cache_gen_fn;
+            self.caches_by_epoch.insert(epoch, cache_gen_fn(epoch as usize));
+        }
+
+        self.recently_accessed_epochs.sort();
+        self.caches_by_epoch.get(&epoch).unwrap()
+    }
+
+    fn get_cache_for_epoch(epoch: usize) -> Vec<u8> {
+        let seed = ethash::get_seedhash(epoch);
+        let cache_size = ethash::get_cache_size(epoch);
+        let mut data = vec![0; cache_size];
+        ethash::make_cache(data.as_mut_slice(), seed);
+        data
+    }
+}
+
+pub struct EthashProver {
+    /// A LRU cache of DAG caches
+    dags_cache: Option<EthashCache>,
+}
+
+impl EthashProver {
+    pub fn new() -> Self {
+        Self {
+            dags_cache: None,
+        }
+    }
+
+    pub fn with_hashimoto_light(max_cache_entries: usize) -> Self {
+        Self {
+            dags_cache: Some(EthashCache::new(max_cache_entries)),
+        }
+    }
+
+    fn dag_merkle_root(&self, epoch: u64) -> Option<H128> {
+        DAGS_MERKLE_ROOTS.get((epoch - DAGS_START_EPOCH) as usize)
+            .map(|x| H128::from(x))
+    }
+
+    // Adapted fro https://github.com/near/rainbow-bridge/blob/3fcdfbc6c0011f0e1507956a81c820616fb963b4/contracts/near/eth-client/src/lib.rs#L363
+    pub fn hashimoto_merkle(
+        &self,
+        header_hash: H256,
+        nonce: H64,
+        header_number: u64,
+        nodes: &[DoubleNodeWithMerkleProof],
+    ) -> Result<(H256, H256), &'static str> {
+        // Check that we have the expected number of nodes with proofs
+        const MIXHASHES: usize = MIX_BYTES / HASH_BYTES;
+        if nodes.len() != MIXHASHES * ACCESSES / 2 {
+            return Err("Incorrect number of nodes");
```

Comment:
Rather than using strings as the error type, can we introduce an enum like this:

```rust
#[derive(Debug, Clone)]
enum Error {
  UnexpectedNumberOfNodes,
  InvalidProof,
}
```
And then return either `Err(Error::UnexpectedNumberOfNodes)` or `Err(Error::InvalidMerkleProof)` as appropriate.

This makes it easier to pattern the match on possible errors returned by `hashimoto_merkle`.